_posts/en/posts/2016-02-26-zkcp.md

```diff
@@ -0,0 +1,140 @@
+irreversible, potentially crosses multiple jurisdictions, and involves parties whose financial stability is uncertain--meaning that both parties either take a great deal of risk or have to make difficult arrangement. Using a ZKCP avoids the significant transactional costs involved in a
+sale which can otherwise easily go wrong. 
+
+In today's transaction I purchased a solution to a 16x16 Sudoku puzzle for 0.10 BTC  from Sean Bowe, a member of the Zcash team, as part of a demonstration performed live at Financial Cryptography 2016 in Barbados. I played my part in the transaction remotely from California.
+
+The transfer involved two transactions: 
+
+- [FIXME](https://www.blocktrail.com/BTC/tx/FIXME)
+- [FIXME](https://www.blocktrail.com/BTC/tx/FIXME)
+
+Almost all of the engineering work behind this ZKCP implementation was done by Sean Bowe, with support from Pieter Wuille, myself, and Madars Virza.
+
+## Background
+
+I first proposed the ZKCP protocol in 2011 in an [article on the Bitcoin Wiki](https://en.bitcoin.it/wiki/Zero_Knowledge_Contingent_Payment) as an example of how tremendously powerful the existing primitives in Bitcoin Script already were.
+
+### Zero Knowledge Proofs
+
+My ZKCP protocol required as a building block a zero-knowledge proof for arbitrary programs. Many kind of specialized zero-knowledge proofs exist:
+common digital signatures are an example, as are the range proofs in [Confidential Transaction](https://people.xiph.org/~greg/confidential_values.txt).
+
+A zero-knowledge proof for _general computation_ is a cryptographic system which lets a person run an arbitrary program with a mixture of public and secret inputs and prove to others that this specific program accepted the inputs, without revealing anything more about its operation or the secret inputs.
+
+If this seems like impossible magic, for educational purposes I came up with [a very simple but inefficient ZKP system](https://people.xiph.org/~greg/simple_verifyable_execution.txt) that uses nothing fancier than boolean circuits and cryptographic hashes, or see Matthew Green's
+[Graph coloring ZKP example](http://blog.cryptographyengineering.com/2014/11/zero-knowledge-proofs-illustrated-primer.html).
+
+As my initial write-up on ZKCP noted, no such system was readily available in 2011 but they were believed to be possible, especially under specific constraints that would have worked for ZKCP.
+
+In 2012 Gennaro, Gentry, Parno, and Raykova published a paper ("Quadratic Span Programs and Succinct NIZKs without PCPs") which described an especially efficient construction. Since then, several groups have continued to advance this work, creating compilers, performance improvements, and most critically, practical tools like libsnark. The GGPR'12 cryptosystem requires a trusted setup, but for the ZKCP application this is no real limitation since the buyer can perform it. Because of this work, ZKCP can now become a practical tool.
```

Comment:
a link for "Quadratic Span Programs and Succinct NIZKs without PCPs" could be: https://eprint.iacr.org/2012/215